File: ios/TacticalMaps/Util/DecimalInput.swift
import Foundation

/// A scalar input grammar, not a coordinate parser. Dots always mean decimals;
/// the locale's decimal separator is also accepted. Grouping is never stripped.
enum DecimalInput {
    static func parse(_ text: String, locale: Locale = DisplayFormat.currentLocale) -> Double? {
        var value = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let separator = locale.decimalSeparator ?? "."
        if separator != "." { value = value.replacingOccurrences(of: separator, with: ".") }
        let pattern = #"^[+-]?(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)(?:[eE][+-]?[0-9]+)?$"#
        guard value.range(of: pattern, options: .regularExpression) != nil,
              let parsed = Double(value), parsed.isFinite else { return nil }
        return parsed
    }
}

File: android/app/src/main/java/com/tacmap/localization/DecimalInput.kt
package com.tacmap.localization

import java.text.DecimalFormatSymbols
import java.util.Locale

/** Scalar input only: dots always mean decimals. Never strip grouping or parse coordinates here. */
object DecimalInput {
    private val grammar = Regex("[+-]?(?:[0-9]+(?:\\.[0-9]*)?|\\.[0-9]+)(?:[eE][+-]?[0-9]+)?")

    fun parse(text: String, locale: Locale = DisplayFormat.currentLocale): Double? {
        val separator = DecimalFormatSymbols.getInstance(locale).decimalSeparator
        val value = text.trim().replace(separator, '.')
        if (!grammar.matches(value)) return null
        return value.toDoubleOrNull()?.takeIf { it.isFinite() }
    }
}
